src/transaction/types/mint.rs

```diff
@@ -0,0 +1,327 @@
+use crate::transaction::{
+    checkable::Checkable,
+    field::{Outputs, TxPointer as TxPointerField},
+};
+use crate::{CheckError, ConsensusParameters, Output, TxPointer};
+use derivative::Derivative;
+use fuel_types::bytes::{SizedBytes, WORD_SIZE};
+use fuel_types::{Bytes32, Word};
+
+#[cfg(feature = "std")]
+use std::io;
+
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+
+#[cfg(feature = "std")]
+use fuel_types::bytes::{self, Deserializable, SerializableVec};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct MintMetadata {
+    pub id: Bytes32,
+    pub outputs_offset: usize,
+    pub outputs_offset_at: Vec<usize>,
+    pub serialized_size: usize,
+}
+
+#[cfg(feature = "std")]
+impl MintMetadata {
+    fn compute<Tx>(tx: &Tx) -> Self
+    where
+        Tx: crate::UniqueIdentifier,
+        Tx: Outputs,
+        Tx: SizedBytes,
+    {
+        use itertools::Itertools;
+
+        let id = tx.id();
+
+        let mut offset = tx.outputs_offset();
+
+        let outputs_offset = offset;
+        #[cfg(feature = "internals")]
+        assert_eq!(outputs_offset, tx.outputs_offset());
+
+        let outputs_offset_at = tx
+            .outputs()
+            .iter()
+            .map(|output| {
+                let i = offset;
+                offset += output.serialized_size();
+                i
+            })
+            .collect_vec();
+
+        let serialized_size = offset;
+        #[cfg(feature = "internals")]
+        assert_eq!(serialized_size, tx.serialized_size());
+
+        Self {
+            id,
+            outputs_offset,
+            outputs_offset_at,
+            serialized_size,
+        }
+    }
+}
+
+/// The definition of the `Mint` transaction from the specification:
+/// https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#transactionmint
+///
+/// This transaction can be created by the block producer and included in the block only by it.
+#[derive(Default, Debug, Clone, Derivative)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derivative(Eq, PartialEq, Hash)]
+pub struct Mint {
+    /// The location of the transaction in the block.
+    pub(crate) tx_pointer: TxPointer,
```

Comment:
Is this a overloading of usecases for TxPointer? Seems like it's supposed to only apply to inputs.

Reply:
It is kind of a strange case because, the txpointer should be implied from the block producer and it's needed to generate the ID, but it's not included on the mint tx according to the specs. 

Seems like maybe we should update the tx format in the specs to include the txpointer since it's part of the id? And in that case, maybe it should also be included in the canonical serialization?

Reply:
I planned to do this spec change in the scope of https://github.com/FuelLabs/fuel-specs/issues/421. 

Sorry, maybe I didn't describe it well=( Initially, the plan was to include two fields, `block_height` and `tx_index` as fields of `Mint` transaction to make it unique. And if we are okay with it, I will make corresponding modification in the spec=)

`TxPointer` is a representation of those two fields, so it fits well here, but it requires actualization in the specification.

Reply:
Interesting. I wouldn't call this overloading of the TxPointer type; it's a pointer to a transaction in a block after all!

Putting it into the transaction itself would make the transaction larger strictly unnecessarily, but would avoid needing special rules for computing mint transaction IDs vs other transaction IDs. I'm actually inclined to go with this proposal since it means one less rule for computing IDs.

Reply:
resolved: https://github.com/FuelLabs/fuel-specs/pull/423